### ISSUE TITLE
'gm' config generates code which can not build

### ISSUE THREAD

**Issue by VanPuyveldeKevin** [delegateas/XrmDefinitelyTyped#252]
The tool has a useful feature to generate attribute/control mappings. It allows me to do stuff like properly type arrays of attribute names.

However, the feature has long since been mostly unusable since it generates invalid code files for various components (most notably, even a default CRM environment already breaks things).
![image](https://user-images.githubusercontent.com/19588107/151335993-9b74b8b1-5c0f-486d-824e-e9a1bfc556b2.png)
![image](https://user-images.githubusercontent.com/19588107/151337448-dfb67d86-815f-4f18-a498-0715fe0dcc55.png)

As we see in the screenshot, as soon as we hit that point (basically, a guid was used), it breaks as it's no longer valid code.
I get the 'why' of why this happens, but I really would say it's better to 'not' generate those parts that generate invalid code rather than break the whole build.

In Visual Studio, this generally works as long as we don't touch it. However, automated build pipelines just end up failing due to erroneous files.

As such ... I'd say the best approach really remains to not generate invalid code, even if that means those invalid components aren't generated.
What is the DelegateAS team's take on this?

**Comment by misoeli** [comment]
Most important things first: I am a bit interested in figuring out exactly what controls and attributes the issue is caused by. You mention it happen on clean installs, but I'm not seeing that. Could you check with entity and form the issues you screencapped above are from, which generation settings you applied, and preferably also include XrmDT version number.

I agree it doesn't make sense to generate typings with errors. I am, however, not sure whether we actually use that in any of our projects. I haven't heard of the issue internally,  nor encountered it myself (the latter definitely due to the flag not being used).

I'm not really sure where this would be of use, so really curious where you see the relevance of this. You mention arrays of attribute names, but in what scenarios wouldn't it be as good getting the attributes directly from the form context?

That being said, the controls with invalid names should be relatively easy to exclude. We already validate the control names to ensure header and footer controls/attributes are excluded in <v6.0 (so probably not after all, but the check is still left around).

**Comment by VanPuyveldeKevin** [comment]
Aiight, first of all, the offending MS items ...

I've managed to get around the issue at the moment. 
As VS didn't actually fail on this issue (as it generates a JS file from the TS when saving a single TS source file) and the issue was generally centered on the build server I was able to prevent the build server from dying by excluding the relevant items from source control.
Basically ... we've still got our stuff in source control, but I use a .gitignore file at the project's root level to exlude whatever kills the build server. Of course I deleted the breaking d.ts files as well in source control, so the build server doesn't die anymore.

As such, I've exluded these files (.gitignore content):
```
typings/XRM/Form/account/InteractionCentricDashboard/Dashboardvooraccounts.d.ts
typings/XRM/Form/account/InteractionCentricDashboard/test.d.ts
typings/XRM/Form/custom_purchaseorder/InteractionCentricDashboard/inkooporders.d.ts
typings/XRM/Form/contact/InteractionCentricDashboard/Dashboardvoorcontactpersonen.d.ts
typings/XRM/Form/contact/TaskFlowForm/Contactpersoonbijwerken.d.ts
typings/XRM/Form/lead/InteractionCentricDashboard/Dashboardvoorpotentiëleklanten.d.ts
```

Several of these are pretty much on default entities and were not manually created by our team. 
Mostly it's the whole '**`InteractionCentricDashboard`**' & '**`TaskFlowForm`**' stuff that breaks the rules.

**NOTE:** Could be they get added by some solution or other or by checking some kind of checkbox, haven't looked into where the components came from, so maybe they wouldn't be there on a base install w/o any sales/marketing solutions. But in this case this was for an on-premise v9.2 organization.

In any case, these aren't *.d.ts files that we actually use for typing anything we're developing for, so they're safe to exclude.

As for the relevant setting ... the setting that causes the issue to pop up is 'generate mappings'
```xml
<add key="gm" value="true" />
```

I'm also specifying specific entities to include as well as the 'web' key as the most important configs.

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <appSettings>
    <add key="url" value="https://url/XRMServices/2011/Organization.svc" />
    <add key="username" value="" />
    <add key="password" value="" />
    <add key="out" value="../typings/XRM" />
    <add key="solutions" value="" />
    <add key="entities" value="
                                account,
                                ..., />
    <add key="web" value="" />
    <add key="jsLib" value="../src/lib" />
    <add key="mfaAppId" value="" />
    <add key="mfaReturnUrl" value="" />
    <add key="gm" value="true" />
  </appSettings>
</configuration>
```
As for the version ... I tried several, including the very latest one.

----------------------------------------------------------------------
As for the primary reason this setting is so useful ... **arrays & utility methods that are actually type validated**.

Say for a very simple example, that I want to create a method that disables some fields. 
The 'regular' method of doing this would be to just have a long list of 'form.getAttribute("xxx").setDisabled(true/false)' code lines.
However, that'd all be hardcoded, a long way from being useful as a utility method.

```typescript
type EntityForm = Form.new_entity.Main.ExampleEntityForm;
type AttributeVM = keyof Form.new_entity.Main.ExampleEntityForm.AttributeValueMap;

  const AlwaysDisabledAttributeNames: Array<AttributeVM> =
    [
      "new_name",
      "new_field1",
      "new_field2",
      "new_field3",
      "new_field4",
      "new_field5"
    ];
```
And voila ... I now have a type safe list of attributes I can foreach to disable or do whatever I need with rather than repeat a bunch of code 6 times, or 10 or 100 times. The more things need to be done with those attribute names, the more repeated cookie cutter code I save out.

Without `AttributeValueMap`, I cannot do this. Surely I can easily create an array of strings, but none of them would be 'type safe' (e.g. validated by XDT typings). The beauty of this feature is that this array actually is.
![image](https://user-images.githubusercontent.com/19588107/154557434-2948de23-481e-4241-a176-34ef494ee6c0.png)

Say I'd need to make each of those 'required', 'enabled' and 'visible' for example. That'd be 3 lines for each of those attributes.
Let alone that setDisabled is on the control & required on the attribute. You'd end up with a lot of lines of cookie cutter code to do the same thing 6 times with typing validation on the regular form context typings.

Of course this also means I can now create type-safe utility methods that accept d.ts validated strings.

```typescript

function someOnChangeOrWhateverMethodOnTheForm(form: EntityForm): void {
       $utils.setAttributesRequired<AttributeVM>(form, ["field1", "field2", "field3"], true);
    }
```
```typescript
export function setAttributesRequired<TKey extends string>(
        form: Xrm.PageBase<Xrm.AttributeCollectionBase, Xrm.TabCollectionBase, Xrm.ControlCollectionBase>,
        attributes: Array<TKey>,
        required: boolean): void
    {
        let basicPage = <Xrm.BasicPage>form;

        attributes.forEach(attributeName =>
        {
            basicPage.getAttribute(attributeName).setRequiredLevel(required ? "required" : "none");
        });
    }
```
Sure, these utility methods 'could' receive '`Control<any>`' or something like that, but those would still need to be 'gotten' statically on the calling form script, which doesn't quite offer this level of freedom, conciseness & for my taste, readability (since you're basically foreaching on an array of strings, rather than a wall of text doing getAttribute/getControl.xxx calls).

---------------

Anyway ... I put some effort into the above, obviously simplifying stuff a bit, throwing in some generic names. 
I hope that answers your question or was that not the 'relevance' you were inquiring for?

Some feedback or thoughts as to the above would be appreciated though. Whether or not the team considers it a useful application of the feature.

**Comment by misoeli** [comment]
Thank you. We have similar usages, but extracting the attributes directly is probably more clean. It could be interesting to look into at least.

I managed to identify the issue as well. I think VS was acting up when testing, combined with, as you identified, the issue occurring on some of the more specialized forms. At the very least, there's something to go by now. I'll look into it.

**Comment by misoeli** [comment]
Slightly premature close of the issue. Included in v6.0.0. 

Instead of actually trying to fix the name of the controls and attributes, the special forms (InteractionCentricDashboard and TaskFlowForm) are excluded from context generation. I cannot see a scenario where they make any sense having typings for, as there's no way of adding web ressources to them anyways. It would probably be possible to circumvent the current naming convention in these special cases, but I simply cannot see a reason for it.

**Comment by VanPuyveldeKevin** [comment]
Makes sense.
